### DISCUSSION TITLE
Refactoring to remove a boolean flag from context

### DISCUSSION THREAD

**Discussion by DanCouper** [statelyai/xstate#2808] [Q&A]
So I have a machine which deals with auth. Two overall stages: authentication against AWS, then on-device PIN security.

If a user does not have a session and they haven't set a PIN, then they go through authentication then set up a PIN.
If a user does have a session, they go straight to PIN security.

The scenario I'm trying to work with ATM is when a user's session has expired. In that case, they do have a PIN set up, but need to go through authentication. Once they complete authentication, I skip the PIN stage as it's not necessary.

The way I've done this is to set a boolean flag in context, `loginCompleted`, to `false` initially. It's the only flag in the machine. If a user has a session, this stays `false`. If they don't, it flips to `true`. Then when I reach the state where I'm checking if there's a PIN stored on the device, I use a `cond` to check the flag: if `true`, go to authenticated, if `false`, set up a PIN.

This _works_ absolutely fine, but IMO it's awful because it's stupidly painful to test. Without adding context/conditionals to the model tests, which I've thus far completely avoided, I have to create two model tests, one which tests all paths where the authentication flow needs to be undertaken by a user that has a PIN, one where they don't have a PIN.

All of the API/device PIN checks are done outside the machine -- I just send messages when the checks resolve/reject.

Does anyone have any advice on getting rid of flags, because the only alternative I can see here is adding a load of extra states

**Comment by ChrisShank** [comment]
Could you share some more code?

**Reply by DanCouper** [reply]
Ah sorry, got caught up in other work, here we go, I think I've stripped it down enough to be easily readable:

https://stately.ai/viz/9d8f0102-db4c-4997-abb9-e4b5850f702f

_(edit: umm the translation from my code to the visualizer has meant that nothing ever gets assigned to context unless you manually type in the event each time & replacing that with direct assignment in the actions doesn't seem to be doing the job, I'll try and get it at least updating the context)_
